posthog/hogql/database/schema/util/session_where_clause_extractor.py

```diff
@@ -0,0 +1,329 @@
+                        right=node.left,
+                    )
+                )
+
+        return ast.Constant(value=True)
+
+    def visit_arithmetic_operation(self, node: ast.ArithmeticOperation) -> ast.Expr:
+        # don't even try to handle complex logic
+        return ast.Constant(value=True)
+
+    def visit_not(self, node: ast.Not) -> ast.Expr:
+        return ast.Constant(value=True)
+
+    def visit_call(self, node: ast.Call) -> ast.Expr:
+        if node.name == "and":
+            return self.visit_and(ast.And(exprs=node.args))
+        elif node.name == "or":
+            return self.visit_or(ast.Or(exprs=node.args))
+        return ast.Constant(value=True)
+
+    def visit_field(self, node: ast.Field) -> ast.Expr:
+        return ast.Constant(value=True)
+
+    def visit_constant(self, node: ast.Constant) -> ast.Expr:
+        return ast.Constant(value=True)
+
+    def visit_placeholder(self, node: ast.Placeholder) -> ast.Expr:
+        raise Exception()  # this should never happen, as placeholders should be resolved before this runs
+
+    def visit_and(self, node: ast.And) -> ast.Expr:
+        exprs = [self.visit(expr) for expr in node.exprs]
+
+        flattened = []
+        for expr in exprs:
+            if isinstance(expr, ast.And):
+                flattened.extend(expr.exprs)
+            else:
+                flattened.append(expr)
+
+        if any(isinstance(expr, ast.Constant) and expr.value is False for expr in flattened):
+            return ast.Constant(value=False)
+
+        filtered = [expr for expr in flattened if not isinstance(expr, ast.Constant) or expr.value is not True]
+        if len(filtered) == 0:
+            return ast.Constant(value=True)
+        elif len(filtered) == 1:
+            return filtered[0]
+        else:
+            return ast.And(exprs=filtered)
+
+    def visit_or(self, node: ast.Or) -> ast.Expr:
+        exprs = [self.visit(expr) for expr in node.exprs]
+
+        flattened = []
+        for expr in exprs:
+            if isinstance(expr, ast.Or):
+                flattened.extend(expr.exprs)
+            else:
+                flattened.append(expr)
+
+        if any(isinstance(expr, ast.Constant) and expr.value is True for expr in flattened):
+            return ast.Constant(value=True)
+
+        filtered = [expr for expr in flattened if not isinstance(expr, ast.Constant) or expr.value is not False]
+        if len(filtered) == 0:
+            return ast.Constant(value=False)
+        elif len(filtered) == 1:
+            return filtered[0]
+        else:
+            return ast.Or(exprs=filtered)
+
+    def visit_alias(self, node: ast.Alias) -> ast.Expr:
+        return self.visit(node.expr)
+
+
+def is_time_or_interval_constant(expr: ast.Expr) -> bool:
+    return IsTimeOrIntervalConstantVisitor().visit(expr)
+
+
+class IsTimeOrIntervalConstantVisitor(Visitor):
+    def visit_constant(self, node: ast.Constant) -> bool:
+        return True
```

Comment:
Im not so fully sure how this works in Python, but is it not a big no-no to redefine abstracted methods signature? Again, I'm not entirely sure what this class is achieving - is it checking whether some `Expr` results in a `DateTime` type?

Reply:
Hmm. I think ideally the visitor would be parametrised, so I could either have a Visitor[None] or Visitor[bool] or Visitor[ast.Expr]. Can see how possible this is.

Reply:
Done, though the cloning visitor needed to stay as `Any`